File: src/chipschallenge/blockreactions/BlueButtonTo.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.Buttons;
import chipschallenge.SoundPlayer.sounds;

/**
 * Moving to a Blue Button
 */
public class BlueButtonTo extends NoSlipReaction {

    private BlueButtonTo() {
    }
    private static BlueButtonTo mInstance = null;

    public static synchronized BlueButtonTo getInstance() {
        if (mInstance == null) {
            mInstance = new BlueButtonTo();
        }
        return mInstance;
    }

    public void react(Block moving, Block standing) {
        Buttons.blueButtonDown(standing);
        sound().playSound(sounds.BUTTON);
    }

    public boolean canMove(Block moving, Block standing) {
        return true;
    }
}

File: ChipsChallenge.Shared/Blockreactions/BlueButtonTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    public class BlueButtonTo : NoSlipReaction
    {
        private static BlueButtonTo instance;

        private BlueButtonTo()
        {
        }

        public static BlueButtonTo Instance
        {
            get
            {
                lock (typeof(BlueButtonTo))
                {
                    return instance ?? (instance = new BlueButtonTo());
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
            Buttons.BlueButtonDown(standing);
            Sound().Play(Shared.Sound.Button);
        }

        public override bool canMove(Block moving, Block standing)
        {
            return true;
        }
    }
}